### PR TITLE
Add new ClientHellos

### PR DESCRIPTION
**Pull request by hwh33**
Also add faked support for token binding, ALPS, and delegated credentials.

This add new `ClientHelloSpec` based on:
- Firefox 105
- Chrome 106
- Edge 106
- Safari 16.0
- 360 Secure Browser 11.0
- QQ Browser 11.1

Unfortunately, the specs based on Edge 106 and 360 11.0 seem to incompatible with this library.  I haven't had time yet to investigate why.  I added the specs anyway in case they are useful to some people.  Perhaps they will function in limited use cases.  I did not update `HelloEdge_Auto` and `Hello360_Auto` to avoid breaking code dependent on those references.

### CODE REVIEW

**Comment by hwh33** [comment]
There's been some discussion about recent GWF blocking in https://github.com/net4people/bbs/issues/129.  In that thread, there was discussion about TLS fingerprinting and this library was specifically called out as a means to combat this.  I can't say for sure that these latest ClientHellos will combat any fingerprinting the GFW is doing, but they should at least help clients look more like normal browser traffic.

File: u_fingerprinter.go [hunk 1]
```diff
@@ -351,7 +406,7 @@ func (f *Fingerprinter) FingerprintClientHello(data []byte) (*ClientHelloSpec, e
 			} else if f.AllowBluntMimicry {
 				clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &GenericExtension{extension, extData})
 			} else {
-				return nil, fmt.Errorf("unsupported extension %#x", extension)
+				return nil, fmt.Errorf("unsupported extension %d", extension)
```
Comment (hwh33) [inline u_fingerprinter.go:409]: Drive-by: TLS extension IDs are generally referenced as decimal

**Comment by hwh33** [comment]
The diff is a little mangled for `u_parrots.go`.  I didn't modify existing hellos, I only added new ones.

**Comment by gaukas** [comment]
Happy to get this merged. 

But does Edge 106 (also other browsers) use a different fingerprint than one of these existing Chrome fingerprint? I'm under the impression that it is based on Chromium.

**Comment by hwh33** [comment]
> But does Edge 106 (also other browsers) use a different fingerprint than one of these existing Chrome fingerprint? I'm under the impression that it is based on Chromium.

Yeah, that's a good question.  I haven't checked equality of these new hellos against any of the Chrome hellos beyond 83.  I did this work on a fork; I didn't realize some newer hellos had been integrated upstream.  You are correct that Edge is based on Chromium (and 360 and QQ IIRC).

I don't have time to check equality right now, but I can try to do so in the next couple of days.

**Comment by gaukas** [comment]
I will run a check later based on your fork and let you know the result, and we can go from there.

**Comment by gaukas** [comment]
I think `HelloChrome_106` duplicates `HelloChrome_102`. For compatibility consideration, I would recommend not including `HelloChrome_106` for now. 

Otherwise, all good. All ClientHello fingerprints are unique.

File: u_common.go [hunk 2]
```diff
@@ -164,6 +182,7 @@ var (
 	HelloChrome_96   = ClientHelloID{helloChrome, "96", nil}
 	HelloChrome_100  = ClientHelloID{helloChrome, "100", nil}
 	HelloChrome_102  = ClientHelloID{helloChrome, "102", nil}
+	HelloChrome_106  = ClientHelloID{helloChrome, "106", nil}
```
Comment (gaukas) [inline u_common.go:185]: Since `Chrome_106` shares the fingerprint with `Chrome_102`, let's keep `102` and get rid of `106`.

File: u_common.go [hunk 3]
```diff
@@ -146,15 +163,16 @@ var (
...
 	HelloFirefox_55   = ClientHelloID{helloFirefox, "55", nil}
 	HelloFirefox_56   = ClientHelloID{helloFirefox, "56", nil}
 	HelloFirefox_63   = ClientHelloID{helloFirefox, "63", nil}
 	HelloFirefox_65   = ClientHelloID{helloFirefox, "65", nil}
 	HelloFirefox_99   = ClientHelloID{helloFirefox, "99", nil}
 	HelloFirefox_102  = ClientHelloID{helloFirefox, "102", nil}
+	HelloFirefox_105  = ClientHelloID{helloFirefox, "105", nil}
 
-	HelloChrome_Auto = HelloChrome_102
+	HelloChrome_Auto = HelloChrome_106
```
Comment (gaukas) [inline u_common.go:175]: If we are getting rid of `106`, `Chrome_Auto` should fall back to `Chrome_102`

**Review by gaukas — CHANGES_REQUESTED** [review]

**Comment by hwh33** [comment]
Thanks for running those checks @gaukas!  I updated according to your suggestions.

**Review by gaukas — APPROVED** [review]

**Comment by gaukas** [comment]
Thanks @hwh33! 
And just one more thing before I get this merged... If I understand correctly, I think `ApplicationSettingsExtension` duplicates with the newly added `FakeALPSExtension`? Since we are not really implementing any functional support for `ApplicationSettingsExtension`, I think we'd rather keep the `FakeALPSExtension` and get rid of the former one. What do you think?

**Comment by hwh33** [comment]
Oh, I didn't notice `ApplicationSettingsExtension`.  Hmm I agree that `FakeALPSExtension` is a better name since, as you say, there is no functional support for ALPS.  However, removing `ApplicationSettingsExtension` would be a breaking change for users of this library.

I would recommend keeping the existing type and adding a note to indicate that it is a faked extension.  I've done this here, in a separate commit: https://github.com/refraction-networking/utls/pull/122/commits/432a2f0174ebdc4adc04c3a9a9b60218f80e3d49.  What do you think?

I can squash commits before merging or keep both for posterity.  Let me know which you would prefer.

File: common.go [hunk 4]
```diff
@@ -96,7 +96,6 @@ const (
 	extensionSignatureAlgorithmsCert uint16 = 50
 	extensionKeyShare                uint16 = 51
 	extensionNextProtoNeg            uint16 = 13172 // not IANA assigned
-	extensionALPS                    uint16 = 17513
```
Comment (hwh33) [inline common.go:99]: This library generally does not modify the standard lib files.  I removed this in favor of the `fakeExtensionALPS` value I added to `u_common.go`.

Comment (gaukas) [reply common.go:99]: Looks good. Thanks!

**Comment by gaukas** [comment]
I'm think we may instead keep an alias. 
```
type ApplicationSettingsExtension = FakeALPSExtension
```
or the other way. 

I wonder, are there any difference between `FakeALPSExtension` and `ApplicationSettingsExtension` in your last commit, except for their naming?

**Comment by gaukas** [comment]
But I guess that's another issue. How about this: I will get this merged, and we fix that issue with a new PR. 
How does this sound to you?

**Comment by gaukas** [comment]
> They changed in Chrome 98 to only include tls versions 1.3 and 1.2.

Nice catch! I think we could make a change if that is the case... But anyways, in the latest Chrome 102's fingerprint we are only including TLS 1.3 and TLS 1.2 in supported versions.

> encrypted extensions in our branch

Interesting. Does `crypto/tls` implement such? If so we still have a pending upstream synchronization. If not... we always welcome PRs! 

> Google sites will reject uTLS without this implementation.

Interesting observation. I thought Google was rejecting uTLS simply because we didn't implement ALPS extension.

**Comment by blakebyrnes** [comment]
> Nice catch! I think we could make a change if that is the case... But anyways, in the latest Chrome 102's fingerprint we are only including TLS 1.3 and TLS 1.2 in supported versions.



Oh, I totally missed 102 since it's out of order with the rest of the Chromes! My bad.

> Interesting. Does crypto/tls implement such? If so we still have a pending upstream synchronization. If not... we always welcome PRs!

crypto/tls implements handshake messages.. which is where some of this code has to live. My point is that uTLS hasn't created a structure like the u_conn and u_common whereby connection state, and new handshake messages can be added. That means this codebase is going to get very messy for things like Compressed Certificates and Encrypted Extensions as long as there's no pattern. I'm honestly not sure the right pattern here.. I have a little bit of a hard time following the current flow.

>Interesting observation. I thought Google was rejecting uTLS simply because we didn't implement ALPS extension.

Sort of the same thing. ALPS "implementation" is to respond to other handshake messages initiated by the server. It's a movement of "TLS extensions" into an encrypted follow-up handshake message, instead of broadcasting in the clear in the clientHello.
